File: src/main/resources/java/br/on/daed/services/controllers/SiteController.java
package br.on.daed.services.controllers;

import br.on.daed.services.html.HTMLHandling;
import br.on.daed.services.html.portos.TabuaMare;
import br.on.daed.services.pdf.DadosMagneticos;
import java.io.File;
import java.io.FilenameFilter;
import java.io.IOException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.springframework.stereotype.Controller;
import org.springframework.ui.ModelMap;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseBody;

@Controller
@RequestMapping("/")
public class SiteController {

    private String setWebGLTemplate(ModelMap map, String path) {
        map.addAttribute("conteudo", path);
        return "templatewebgl";
    }

    @RequestMapping("/angulo-horario")
    public String coordenadasHorarias(ModelMap map) {
        return setWebGLTemplate(map, "angulo-horario");
    }

    @RequestMapping("/")
    public String index(ModelMap map) {
        return "mainpage";
    }

    @RequestMapping("/obliquidade-da-ecliptica")
    public String ecliptica(ModelMap map) {
        return setWebGLTemplate(map, "obliquidade-da-ecliptica");
    }

    @RequestMapping("/coordenadas-supergalacticas")
    public String coordenadasSupergalacticas(ModelMap map) {
        return setWebGLTemplate(map, "coordenadas-supergalacticas");
    }

    @RequestMapping("/coordenadas-galacticas")
    public String coordenadasGalacticas(ModelMap map) {
        return setWebGLTemplate(map, "coordenadas-galacticas");
    }

    @RequestMapping("/coordenadas-eclipticas")
    public String coordenadasEclipticas(ModelMap map) {
        return setWebGLTemplate(map, "coordenadas-eclipticas");
    }

    @RequestMapping("/coordenadas-horizontais")
    public String coordenadasHorizontais(ModelMap map) {
        return setWebGLTemplate(map, "coordenadas-horizontais");
    }

    @RequestMapping("/coordenadas-equatoriais")
    public String coordenadasEquatoriais(ModelMap map) {
        return setWebGLTemplate(map, "coordenadas-equatoriais");
    }

    @RequestMapping("/movimentos-da-terra")
    public String movimentosTerra(ModelMap map) {
        return setWebGLTemplate(map, "movimentos-da-terra");
    }

    @RequestMapping("/calendario-gregoriano")
    public String dataGregoriana(ModelMap map) {
        return setWebGLTemplate(map, "calendario-gregoriano");
    }

    @RequestMapping("/data-juliana")
    public String dataJuliana(ModelMap map) {
        return setWebGLTemplate(map, "data-juliana");
    }

    @RequestMapping("/posicao-sol")
    public String posicaoSol(ModelMap map) {
        return setWebGLTemplate(map, "posicao-sol");
    }

    @RequestMapping("/posicao-lua")
    public String posicaoLua(ModelMap map) {
        return setWebGLTemplate(map, "posicao-lua");
    }

    @RequestMapping("/eclipses")
    public String eclipseSolar(ModelMap map) {
        return setWebGLTemplate(map, "eclipses");
    }

    @RequestMapping("/equacao-de-kepler")
    public String equacaoKepler(ModelMap map) {
        return setWebGLTemplate(map, "equacao-de-kepler");
    }

	@RequestMapping("/satelites-jupiter")
    public String satelitesJupiter(ModelMap map) {
        return setWebGLTemplate(map, "satelites-jupiter");
    }
	
    @RequestMapping("/linhas-de-forca")
    public String linhasDeForca(ModelMap map) {
        return setWebGLTemplate(map, "linhas-de-forca");
    }

    @RequestMapping("/magnetismo-terrestre")
    public String magnetismoTerrestre(ModelMap map) {
        return setWebGLTemplate(map, "magnetismo-terrestre");
    }
    
    private final String GRAFICO_DEFAULT = "ead2015";

    @RequestMapping("/grafico-globo")
    public String graficoGlobo(ModelMap map, @RequestParam(value = "data", required = false) String conteudo) {
        map.addAttribute("conteudo", conteudo == null ? GRAFICO_DEFAULT : conteudo);
        return "webgl/grafico-globo/grafico-globo";
    }

    @RequestMapping("/holo-grafico-globo")
    public String holoPiramide(ModelMap map, @RequestParam(value = "data", required = false) String conteudo) {
        map.addAttribute("conteudo", conteudo == null ? GRAFICO_DEFAULT : conteudo);
        return "webgl/holo-piramide/grafico-globo";
    }

    @RequestMapping("/holo-grafico-globo-2")
    public String holoPiramide2(ModelMap map, @RequestParam(value = "data", required = false) String conteudo) {
        map.addAttribute("conteudo", conteudo == null ? GRAFICO_DEFAULT : conteudo);
        return "webgl/holo-piramide/grafico-globo-2";
    }

    @RequestMapping("/tabua-mares")
    public @ResponseBody
    Object tabuaMares(ModelMap map) {
        return HTMLHandling.getMareOptions();
    }

    @RequestMapping("/mares")
    public String mares(ModelMap map) {
		map.addAttribute("maxAnoTabuas", HTMLHandling.getMareYear());
        return setWebGLTemplate(map, "mares");
    }

    @RequestMapping("/tabua-mares/{tabuaMares}")
    public @ResponseBody
    TabuaMare tabuaMares(@PathVariable String tabuaMares) {
        return HTMLHandling.getTabuaMare(tabuaMares);
    }

    private File[] getArquivosOrbitas(HttpServletRequest request) {
        String path = request.getServletContext().getRealPath("/WEB-INF/classes/static/lib/on-daed-js/orbitas");

        File folder = new File(path);

        File[] files = folder.listFiles(new FilenameFilter() {

            @Override
            public boolean accept(File dir, String name) {
                return name.endsWith(".json") && !name.equals("referencia.json");
            }

        });

        return files;
    }

    @RequestMapping("/orbitas")
    public void orbitas(HttpServletRequest request, HttpServletResponse response) throws IOException {
        File[] arquivosOrbitas = getArquivosOrbitas(request);

        if (arquivosOrbitas.length > 0) {
            response.sendRedirect("orbitas-" + arquivosOrbitas[0].getName().replace(".json", ""));
        }
    }

    @RequestMapping("/orbitas-{orbita}")
    public String orbitas(HttpServletRequest request, ModelMap map, @PathVariable String orbita) {
        String filename = orbita + ".json";

        File[] arquivosOrbitas = getArquivosOrbitas(request);

        String arquivoEscolhido = null;

        for (int i = 0; i < arquivosOrbitas.length; i++) {
            if (arquivosOrbitas[i].getName().equals(filename)) {
                arquivoEscolhido = filename;
            }
        }

        if (arquivoEscolhido == null) {
            arquivoEscolhido = arquivosOrbitas[0].getName();
        }

        map.addAttribute("arquivoDados", arquivoEscolhido);

        return setWebGLTemplate(map, "orbitas");
    }

    @RequestMapping(value = "/pdf/dados-magneticos.pdf")
    @ResponseBody
    public byte[] dadosMagneticos(HttpServletResponse response, @RequestParam(value = "ano") String ano, @RequestParam(value = "tipo") String tipo) throws UnsupportedOperationException, IOException, InterruptedException {
        response.setContentType("application/pdf");
        response.setHeader("Content-Disposition", "attachment; filename=dados-magneticos.pdf");
        return DadosMagneticos.gerarPDF(ano, tipo);
    }

}

File: dotnet/Astro.Web/Controllers/SiteController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using Astro.Domain.Services;
using Astro.Web.Services;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.FileProviders;
using Microsoft.Extensions.Logging;

namespace Astro.Web.Controllers;

[ApiExplorerSettings(IgnoreApi = true)]
[Route("")]
public class SiteController : Controller
{
    private const string WebGlViewName = "TemplateWebGl";
    private const string WebGlContentKey = "conteudo";
    private const string OrbitasViewName = "orbitas";
    private const string GraficoDefault = "ead2015";

    private readonly ITideTableService _tideTables;
    private readonly IMagneticPdfService _magneticPdf;
    private readonly IFileProvider _webRootFiles;
    private readonly ILogger<SiteController> _logger;
    private readonly IFileProvider _templateFiles;
    private readonly TemplateResolver _resolver;
    private readonly string? _creditosFragment;

    public SiteController(
        ITideTableService tideTables,
        IMagneticPdfService magneticPdf,
        IWebHostEnvironment environment,
        ILogger<SiteController> logger)
    {
        _tideTables = tideTables;
        _magneticPdf = magneticPdf;
        _logger = logger;
        _webRootFiles = environment.WebRootFileProvider;
        var templateCandidates = new[]
        {
            Path.Combine(environment.ContentRootPath, "LegacyTemplates"),
            Path.Combine(environment.ContentRootPath, "..", "..", "src", "main", "resources", "templates")
        };

        var templatesPath = templateCandidates.FirstOrDefault(Directory.Exists);
        _templateFiles = templatesPath is not null
            ? new PhysicalFileProvider(templatesPath)
            : new NullFileProvider();
        _resolver = new TemplateResolver(_templateFiles);
        _creditosFragment = _resolver.ResolveExpression("creditos :: creditos");
        if (!string.IsNullOrWhiteSpace(_creditosFragment))
        {
            _logger.LogDebug("Credit fragment sample: {Snippet}", _creditosFragment);
        }
    }

    [HttpGet("")]
    public IActionResult Index()
    {
        var html = _resolver.Resolve("mainpage.html");
        if (string.IsNullOrWhiteSpace(html))
        {
            return NotFound();
        }
        return Content(html, "text/html; charset=utf-8");
    }

    [HttpGet("angulo-horario")]
    public IActionResult CoordenadasHorarias() => WebGl("angulo-horario");

    [HttpGet("obliquidade-da-ecliptica")]
    public IActionResult ObliquidadeDaEcliptica() => WebGl("obliquidade-da-ecliptica");

    [HttpGet("coordenadas-supergalacticas")]
    public IActionResult CoordenadasSupergalacticas() => WebGl("coordenadas-supergalacticas");

    [HttpGet("coordenadas-galacticas")]
    public IActionResult CoordenadasGalacticas() => WebGl("coordenadas-galacticas");

    [HttpGet("coordenadas-eclipticas")]
    public IActionResult CoordenadasEclipticas() => WebGl("coordenadas-eclipticas");

    [HttpGet("coordenadas-horizontais")]
    public IActionResult CoordenadasHorizontais() => WebGl("coordenadas-horizontais");

    [HttpGet("coordenadas-equatoriais")]
    public IActionResult CoordenadasEquatoriais() => WebGl("coordenadas-equatoriais");

    [HttpGet("movimentos-da-terra")]
    public IActionResult MovimentosDaTerra() => WebGl("movimentos-da-terra");

    [HttpGet("calendario-gregoriano")]
    public IActionResult CalendarioGregoriano() => WebGl("calendario-gregoriano");

    [HttpGet("data-juliana")]
    public IActionResult DataJuliana() => WebGl("data-juliana");

    [HttpGet("posicao-sol")]
    public IActionResult PosicaoSol() => WebGl("posicao-sol");

    [HttpGet("posicao-lua")]
    public IActionResult PosicaoLua() => WebGl("posicao-lua");

    [HttpGet("eclipses")]
    public IActionResult Eclipses() => WebGl("eclipses");

    [HttpGet("equacao-de-kepler")]
    public IActionResult EquacaoDeKepler() => WebGl("equacao-de-kepler");

    [HttpGet("satelites-jupiter")]
    public IActionResult SatelitesJupiter() => WebGl("satelites-jupiter");

    [HttpGet("old-satelites-jupiter")]
    public IActionResult OldSatelitesJupiter() => WebGl("old-satelites-jupiter");

    [HttpGet("linhas-de-forca")]
    public IActionResult LinhasDeForca() => WebGl("linhas-de-forca");

    [HttpGet("magnetismo-terrestre")]
    public IActionResult MagnetismoTerrestre() => WebGl("magnetismo-terrestre");

    [HttpGet("grafico-globo")]
    public IActionResult GraficoGlobo([FromQuery(Name = "data")] string? conteudo)
    {
        ViewData[WebGlContentKey] = string.IsNullOrWhiteSpace(conteudo) ? GraficoDefault : conteudo;
        return View("webgl/grafico-globo/grafico-globo");
    }

    [HttpGet("holo-grafico-globo")]
    public IActionResult HoloGraficoGlobo([FromQuery(Name = "data")] string? conteudo)
    {
        ViewData[WebGlContentKey] = string.IsNullOrWhiteSpace(conteudo) ? GraficoDefault : conteudo;
        return View("webgl/holo-piramide/grafico-globo");
    }

    [HttpGet("holo-grafico-globo-2")]
    public IActionResult HoloGraficoGlobo2([FromQuery(Name = "data")] string? conteudo)
    {
        ViewData[WebGlContentKey] = string.IsNullOrWhiteSpace(conteudo) ? GraficoDefault : conteudo;
        return View("webgl/holo-piramide/grafico-globo-2");
    }

    [HttpGet("tabua-mares")]
    public async Task<IActionResult> TabuaMaresAsync(CancellationToken cancellationToken)
    {
        try
        {
            var options = await _tideTables.GetPortsAsync(cancellationToken);
            if (options is null)
            {
                return StatusCode(StatusCodes.Status503ServiceUnavailable);
            }

            return Ok(options);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to retrieve tide table options.");
            return StatusCode(StatusCodes.Status503ServiceUnavailable);
        }
    }

    [HttpGet("mares")]
    public async Task<IActionResult> MaresAsync(CancellationToken cancellationToken)
    {
        int? maxYear = null;
        try
        {
            maxYear = await _tideTables.GetLatestYearAsync(cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to resolve tide table latest year.");
        }

        if (maxYear.HasValue)
        {
            ViewData["maxAnoTabuas"] = maxYear;
        }

        return WebGl("mares");
    }

    [HttpGet("tabua-mares/{tabuaMares}")]
    public async Task<IActionResult> TabuaMaresAsync(string tabuaMares, CancellationToken cancellationToken)
    {
        try
        {
            var tideTable = await _tideTables.GetTideTableAsync(tabuaMares, cancellationToken);
            if (tideTable is null)
            {
                return StatusCode(StatusCodes.Status503ServiceUnavailable);
            }

            return Ok(tideTable);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to retrieve tide table {Tabua}.", tabuaMares);
            return StatusCode(StatusCodes.Status503ServiceUnavailable);
        }
    }

    [HttpGet("orbitas")]
    public IActionResult Orbitas()
    {
        var orbitNames = LoadOrbitFileNames();
        var target = orbitNames.FirstOrDefault();

        if (string.IsNullOrEmpty(target))
        {
            _logger.LogWarning("No orbit JSON files were found under wwwroot/lib/on-daed-js/orbitas.");
            return NotFound();
        }

        return Redirect($"orbitas-{target}");
    }

    [HttpGet("orbitas-{orbita}")]
    public IActionResult Orbitas(string orbita)
    {
        var orbitNames = LoadOrbitFileNames();
        var selected = orbitNames.FirstOrDefault(name =>
            string.Equals(name, orbita, StringComparison.OrdinalIgnoreCase));

        if (string.IsNullOrEmpty(selected))
        {
            selected = orbitNames.FirstOrDefault();
        }

        if (string.IsNullOrEmpty(selected))
        {
            _logger.LogWarning("No orbit JSON files were found when resolving orbit '{Orbit}'.", orbita);
            return NotFound();
        }

        ViewData["arquivoDados"] = $"{selected}.json";
        return WebGl(OrbitasViewName);
    }

    [HttpGet("pdf/dados-magneticos.pdf")]
    public async Task<IActionResult> DadosMagneticosPdfAsync(
        [FromQuery(Name = "ano")] string ano,
        [FromQuery(Name = "tipo")] string tipo,
        CancellationToken cancellationToken)
    {
        try
        {
            var payload = await _magneticPdf.GeneratePdfAsync(ano, tipo, cancellationToken);
            const string contentType = "application/pdf";
            const string fileName = "dados-magneticos.pdf";
            return File(payload, contentType, fileName);
        }
        catch (ArgumentOutOfRangeException ex)
        {
            _logger.LogWarning(ex, "Invalid request for magnetic PDF with year {Ano} and type {Tipo}.", ano, tipo);
            return BadRequest(new { error = ex.Message });
        }
        catch (FileNotFoundException ex)
        {
            _logger.LogError(ex, "Magnetic PDF output not found.");
            return StatusCode(StatusCodes.Status500InternalServerError);
        }
        catch (InvalidOperationException ex)
        {
            _logger.LogError(ex, "Magnetic PDF generation failed.");
            return StatusCode(StatusCodes.Status500InternalServerError);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Unexpected error generating magnetic PDF.");
            return StatusCode(StatusCodes.Status500InternalServerError);
        }
    }

    private IActionResult WebGl(string contentKey)
    {
        ViewData[WebGlContentKey] = contentKey;
        ViewData["WebGlCss"] = ApplyInlineExpressions(LoadTemplateFragment($"webgl/{contentKey}/css"));
        ViewData["WebGlAjuda"] = ApplyInlineExpressions(LoadTemplateFragment($"webgl/{contentKey}/ajuda"));
        ViewData["WebGlSobre"] = ApplyInlineExpressions(LoadTemplateFragment($"webgl/{contentKey}/sobre"));
        ViewData["WebGlConteudo"] = ApplyInlineExpressions(LoadTemplateFragment($"webgl/{contentKey}/conteudo"));
        ViewData["WebGlCreditos"] = ApplyInlineExpressions(_creditosFragment ?? string.Empty);
        return View(WebGlViewName);
    }

    private IReadOnlyList<string> LoadOrbitFileNames()
    {
        var directory = _webRootFiles.GetDirectoryContents("lib/on-daed-js/orbitas");
        if (!directory.Exists)
        {
            return Array.Empty<string>();
        }

        return directory
            .Where(file => !file.IsDirectory && file.Name.EndsWith(".json", StringComparison.OrdinalIgnoreCase))
            .Select(file => Path.GetFileNameWithoutExtension(file.Name))
            .Where(name => !string.Equals(name, "referencia", StringComparison.OrdinalIgnoreCase))
            .OrderBy(name => name, StringComparer.OrdinalIgnoreCase)
            .ToArray();
    }

    private string? LoadTemplateFragment(string relativePath)
    {
        if (relativePath.StartsWith("/", StringComparison.Ordinal))
        {
            relativePath = relativePath.TrimStart('/');
        }

        return _resolver.Resolve(relativePath);
    }

    private string ApplyInlineExpressions(string? content)
    {
        if (string.IsNullOrEmpty(content))
        {
            return string.Empty;
        }

        return Regex.Replace(content, @"\[\[\$\{([^}]+)\}\]\]", match =>
        {
            var key = match.Groups[1].Value.Trim();
            if (ViewData.TryGetValue(key, out var value) && value is not null)
            {
                return value.ToString() ?? string.Empty;
            }

            return string.Empty;
        });
    }
}
